---
title: "Loading and preprocessing of the CART data"
author: "Konstantinos Charitidis"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    toc: true
    toc_depth: '3'
    df_print: paged
  bookdown::html_document2:
    toc: true
    toc_depth: 3
    toc_float:
      collapse: false
    number_sections: true
    df_print: paged
    theme: readable
    code_folding: hide
    self_contained: true
    highlight: tango
---

```{r setup, include = FALSE}

knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE, warning = FALSE)
options(max.print = 200)

```

```{r Libraries, include = FALSE}

source(file = "~/P_lab/Spatial_CART/Scripts/Other/My_Functions.R")
source(file = "~/P_lab/Spatial_CART/Scripts/Other/CondenseXY.R")
Init_project("~/P_lab/Spatial_CART/")
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tiff)
# library(Matrix)
library(patchwork)
library(stringr)

```

```{r Load the files}

My_Seurat_objs <- Load_My_Rds(file.path(main_dir, "Data"))

```

```{r Merge the objects into one}

My_Seurat_objs_clean <- lapply(My_Seurat_objs, function(obj) {
  obj <- DietSeurat(obj, 
                    assays = c("RNA", "falsecode", "negprobes"), 
                    dimreducs = NULL,  
                    graphs = NULL,
                    misc = FALSE)
  
  obj@images <- list()
  obj
  
} )

## Function works after downloading SeuratObject fix with devtools, throws an sf.object error ##
## https://github.com/satijalab/seurat/issues/10111 ##

Seurat_merged <- merge(x = My_Seurat_objs_clean[[1]], 
                       y = My_Seurat_objs_clean[-1],
                       add.cell.ids = names(My_Seurat_objs_clean),
                       project = "CosMx_CART")

```


```{r Create the sample id columns}

Sample_meta <- Seurat_merged@meta.data %>% 
  mutate(Slide_num = as.integer(str_extract(Run_Tissue_name, "(?<=Slide)\\d+")), 
         Slide = paste0("Slide_", Slide_num), 
         Sample = case_when(Slide_num == 1 & fov >=  1 & fov <= 28 ~ 4,
                            Slide_num == 1 & fov >= 29 & fov <= 41 ~ 3,
                            Slide_num == 1 & fov >= 42 & fov <= 54 ~ 2,
                            Slide_num == 1 & fov >= 55 & fov <= 71 ~ 1,
                            Slide_num == 2 ~ 5, 
                            Slide_num == 3 ~ 6, 
                            Slide_num == 4 ~ 7, 
                            Slide_num == 31 ~ 8, 
                            Slide_num == 32 ~ 9,
                            TRUE ~ NA), 
         Group = case_when(Slide_num == 1 ~ "CART",
                           Slide_num == 2 ~ "Control", 
                           Slide_num %in% c(3, 4, 31, 32) ~ "Conventional", 
                           TRUE ~ NA_character_), 
         Group = factor(Group, levels = c("CART", "Control", "Conventional")), 
         Biopsy = case_when(Sample == 1 ~ "T0", 
                            Sample == 2 ~ "T12", 
                            Sample == 3 ~ "T0", 
                            Sample == 4 ~ "T12", 
                            Sample == 5 ~ "T0", 
                            Sample == 6 ~ "T0", 
                            Sample == 7 ~ "T12", 
                            Sample == 8 ~ "T0", 
                            Sample == 9 ~ "T12", 
                            TRUE ~ NA))

```


```{r Condense the coordinates and apply to the metadata}

XY <- Sample_meta %>%
  select(cell_ID, fov, x_FOV_px, y_FOV_px, x_slide_mm, y_slide_mm, Slide, Sample, Group)

FOV_vec <- XY$fov
Tissue_vec <- factor(paste(XY$Slide, XY$Sample, sep = "_"))

XY <- XY %>%
  select(x_slide_mm, y_slide_mm) %>%
  transmute(x = x_slide_mm,
            y = y_slide_mm) %>%
  as.matrix()

XY_cond <- condenseXY(xy = XY,
                      fov = FOV_vec,
                      tissue = Tissue_vec,
                      condensefovs = TRUE,
                      condensetissues = TRUE,
                      eps = 1.5,
                      mindist = 0.1,
                      fovbuffer = 0.2,
                      tissueorder = NULL,
                      tissuebuffer = 0,
                      widthheightratio = 4/3)

# Build a tibble with coords + cluster labels
Sample_meta <- as.data.frame(XY_cond, .name_repair = "minimal") %>%
  setNames(c("x_cond", "y_cond")) %>%
  rownames_to_column(var = "cell_ID") %>%
  inner_join(Sample_meta %>% mutate(cell_ID = rownames(Sample_meta)), by = "cell_ID") %>% 
  tibble::column_to_rownames("cell_ID") %>%
  as.data.frame()

Seurat_merged@meta.data <- Sample_meta
Seurat_merged@meta.data$cell_ID <- rownames(Seurat_merged@meta.data)

```

```{r Condense vol 2}

# xy_cond_within_group <- condenseXY_within_group(df = Sample_meta,
#                                                 xcol = "x_slide_mm",
#                                                 ycol = "y_slide_mm",
#                                                 groupcol = "Group",
#                                                 samplecol = "Sample",
#                                                 fovcol = "fov",
#                                                 condensefovs = TRUE,
#                                                 condensetissues = TRUE,
#                                                 eps = 1.5,
#                                                 mindist = 0.1,
#                                                 fovbuffer = 0.2,
#                                                 tissueorder = NULL,
#                                                 tissuebuffer = 0,          
#                                                 widthheightratio = 4/3)
# 
# xy_cond_horiz <- xy_cond_groups_horizontal(meta = Sample_meta,
#                                            xy_cond = xy_cond_within_group,
#                                            groupcol = "Group",
#                                            group_order = c("CART", "Control", "Conventional"),
#                                            gap = 0.5)
# 
# 
# Sample_meta <- xy_cond_horiz %>% 
#   inner_join(y = xy_cond_within_group, by = c("cell_ID", "Group")) %>% 
#   inner_join(y = Sample_meta, by = c("cell_ID", "Group"))
# 
# Sample_meta %>%
#   mutate(Sample = factor(Sample)) %>%
#   ggplot(., aes(x = x_cond, y = y_cond, colour = Sample)) +
#   geom_point(size = 0.1, alpha = 0.6) +
#   coord_equal() +
#   facet_wrap(~ Slide_num) +
#   theme_void() +
#   guides(colour = guide_legend(override.aes = list(size = 2))) +
#   labs(colour = "Sample", title = "FOVs by sample")
# 
# Seurat_merged@meta.data <- Sample_meta
# Seurat_merged@meta.data$cell_ID <- Cells(Seurat_merged)
# Seurat_merged@meta.data$cell_ID <- Cells(Seurat_merged)


```

```{r Lets get a look at the FOVs before the anotation}

Sample_meta %>%
  mutate(Sample = factor(Sample)) %>%
  transform(., x_px = x_slide_mm * 0.18,
               y_px = y_slide_mm * 0.18) %>% 
  ggplot(., aes(x = x_cond, y = y_cond, colour = Group)) +
  # ggplot(., aes(x = x_px, y = y_px, colour = Sample)) +
  geom_point(size = 0.1, alpha = 0.6) +
  coord_equal() +
  # facet_wrap(~ Group) +
  theme_void() +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  labs(colour = "Sample", title = "FOVs by sample")

```



```{r Create the Seurat Object}

## Load the TIFF for the slides ##
Tissue_bg <- readTIFF(source = file.path(main_dir, "Data/TIFFs", "MCF-ASCIST-1250-Slide-CART (1979 8769 6063 4322).tiff"), 
                      native = FALSE)

## Convert to raster ##
Tissue_raster <- as.raster(Tissue_bg)

## Add to the object ##
Seurat_merged@misc$background <- Tissue_raster

## Get the cell and gene count pre filtering ##
add_metric(Se_object = Seurat_merged, 
           step = "Pre-filtering", 
           note = "No filtering done yet")

log_info(note = "Data before the filtering",
         "Original number of cells: %d \nOriginal number of genes %d",
         ncol(Seurat_merged), nrow(Seurat_merged))

SaveSeuratRds(object = Seurat_merged, file = file.path(main_dir, "Data", "Edited_data", "Seurat_merged_temp.Rds"))

```

```{r Remove all objects aside from the seurat object}

keep <- c("Seurat_merged", "log_file", "metrics", "qc_dir", "main_dir", "My_Functions.R")
rm(list = setdiff(ls(envir = .GlobalEnv), keep), envir = .GlobalEnv)
gc()
source(file = "~/P_lab/Spatial_CART/Scripts/Other/My_Functions.R")

```

```{r Seurat quality control}

vln_p1 <- VlnPlot(Seurat_merged, features = c("nFeature_RNA", "nCount_RNA"), 
                  ncol = 2, pt.size = 0.001, alpha = 0.1) +
  plot_annotation(title = "CosMx QC before filtering: Gene and total transcript number per cell")

## Get total counts for each cell ##
Seurat_merged$Total_Counts  <- Matrix::colSums(GetAssayData(Seurat_merged, "RNA"))
## Get number of features/genes for each cell ##
Seurat_merged$Num_Features  <- Matrix::colSums(GetAssayData(Seurat_merged, "RNA") > 0)

## Get a visual understanding of their distributions ##
vln_p2 <- VlnPlot(Seurat_merged, features = c("Total_Counts","Num_Features"), 
        log = TRUE, ncol = 2, pt.size = 0.001, alpha = 0.1) +
  plot_annotation(title = "CosMx QC before filtering: Total counts and number of expressed genes per cell")

g_p1 <- ggplot(Seurat_merged@meta.data, aes(Total_Counts, Num_Features)) + 
  geom_point(alpha = 0.3) + 
  labs(title = "Gene complexity vs depth",
       x = "Total transcripts per cell (Total_Counts)",
       y = "Detected genes per cell (Num_Features)")

ggsave(filename = file.path(qc_dir, "/Pre_Filtering/Vln_nFeature_RNA_nCount_RNA.png"),
       plot = vln_p1, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "/Pre_Filtering/Vln_Total_Counts_Num_Features.png"),
       plot = vln_p2, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "/Pre_Filtering/Scatter_Total_Counts_Num_Features.png"),
       plot = g_p1, width = 9, height = 4, dpi = 300)

## Save Rds object to continue from here next session ##
SaveSeuratRds(object = Seurat_merged, file = file.path(main_dir, "Data", "Edited_data", "Seurat_merged.Rds"))

```

```{r Clear space}

# keep <- c("Seurat_merged", "log_file", "metrics", "qc_dir", "main_dir")
# rm(list = setdiff(ls(envir = .GlobalEnv), keep), envir = .GlobalEnv)
# gc()

```

```{r Step 1. Filtering based on genes and cells}

# Seurat_merged <- readRDS(file = file.path(main_dir, "Data", "Edited_data", "Seurat_merged.Rds"))

Meta <- Seurat_merged@meta.data

## Filter based on geometric means of the areas ##
Meta <- Meta %>%
  rownames_to_column("cell_ID") %>%
  ## Get the cell area sizes ##
  ## Calculate the geometric mean of the areas ##
  ## Get the thresholds for each sample ##
  mutate(Sample = as.factor(Sample)) %>%
  group_by(Sample) %>%
  mutate(Area_um2 = as.numeric(Area.um2),
         Geom_mean_area = exp(mean(log(Area_um2[is.finite(Area_um2) & Area_um2 > 0]), na.rm=TRUE)),
         Threshold_area = 5 * Geom_mean_area) %>%
  ungroup() 

## Filter cells with less than 30 transcripts ##
keep_cells <- Meta %>% 
  mutate(prob_insitu = as.numeric(RNA_Export_Cell.Typing.InSituType.1_1_posterior_probability),
         Area_um2 = Area_um2_values) %>% 
  filter(
    ## Filter cells with area less or equal to the threshold ##
    is.finite(Area_um2), Area_um2 > 0, Area_um2 <= Threshold_area,
    ## Filter cells with probability less than 0.8 ##
    # prob_insitu > 0.80,
    ## Filter cells with less than 50 features ##
    nFeature_RNA > 50,
    ## Filter cells with more than 2000 features ##
    nFeature_RNA < 2000) %>% 
  pull(cell_ID)

## Subset based on the cell filtering ##
Seurat_filtered <- subset(Seurat_merged, cells = keep_cells)

## Select genes that are transcribed in at least 1 cell ##
## GetAssayData retrieves the counts (Do this in the filtered object )##
keep_genes <- GetAssayData(Seurat_filtered, assay = "RNA", layer = "counts") %>%
  ## rowSums checks if they are transcribed ##
  { rowSums(. > 0) } %>% 
  ## Conversion of the vector to a tibble/df ##
  enframe(name = "Gene", value = "nonzero") %>%
  ## Keeps the expressed genes #
  filter(nonzero > 0) %>%
  ## Get the gene names only to work with subset() ##
  pull(Gene)

## Filter the RNA assay (only) based on the genes from the above command ##
Counts_assay <- Seurat_filtered[["RNA"]]
Counts_assay <- subset(Counts_assay, features = keep_genes)
Seurat_filtered[["RNA"]] <- Counts_assay

cat(sprintf("Kept %d cells, %d genes\n", 
            ncol(Seurat_filtered), nrow(Seurat_filtered)))

## Save Rds object to continue from here next session ##
SaveSeuratRds(object = Seurat_filtered, file = file.path(main_dir, "Data", "Edited_data", "Seurat_filtered_step_1.Rds"))

```

```{r Clear space}

# keep <- c("Seurat_filtered", "log_file", "metrics", "qc_dir", "main_dir")
# rm(list = setdiff(ls(envir = .GlobalEnv), keep), envir = .GlobalEnv)
# gc()

```

```{r Step 2. Visual QC - Mitochondrial counds}

Counts_filtered <- GetAssayData(Seurat_filtered, layer = "counts")
Mitochondrial_gene_num <- grep("^MT-", rownames(Counts_filtered))

Seurat_filtered$percent.mt <- Matrix::colSums(Counts_filtered[Mitochondrial_gene_num, , drop = FALSE]) /
  Matrix::colSums(Counts_filtered) * 100

paste0("Mitochondrial genes detected: ", length(Mitochondrial_gene_num))

Seurat_filtered$Total_Counts <- Matrix::colSums(Counts_filtered)
Seurat_filtered$Num_Features <- Matrix::colSums(Counts_filtered > 0)

vln_p3 <- VlnPlot(Seurat_filtered, features = c("Total_Counts", "Num_Features", "percent.mt"), 
                  pt.size = 0, ncol = 3) +
  plot_annotation(title = "CosMx QC after filtering (Step 2)")

feat_p1 <- FeatureScatter(Seurat_filtered, feature1 = "Total_Counts", feature2 = "Num_Features")  +
  plot_annotation(title = "CosMx QC after filtering (Step 2): Gene complexity vs depth")

feat_p2 <- FeatureScatter(Seurat_filtered, feature1 = "Total_Counts", feature2 = "percent.mt") +
  plot_annotation(title = "CosMx QC after filtering (Step 2): Mitochondrial counts")

Seurat_filtered@meta.data %>%
  group_by(fov) %>% ## Area most likely contains the FOV information, needs adjustment ##
  summarise(Cells = n(), Median_Counts = median(Total_Counts, na.rm = TRUE),
            Median_Features = median(Num_Features, na.rm = TRUE),
            Median_MT = if ("percent.mt" %in% names(cur_data_all())) 
              median(percent.mt, na.rm = TRUE) else NA_real_) %>%
  arrange(Median_Counts) %>%
  print(n = 40)

ggsave(filename = file.path(qc_dir, "Post_Filtering/Vln_Total_Counts_Features_MT_F_step_2.png"),
       plot = vln_p3, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "Post_Filtering/Scatter_Total_Counts_Features_F_step_2.png"),
       plot = feat_p1, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "Post_Filtering/Scatter_MT_Counts_F_step_2.png"),
       plot = feat_p2, width = 9, height = 4, dpi = 300)

# add_metric(Se_object = Seurat_filtered, 
#            step = "Filtering: Step 2", 
#            note = "Filtering based on MT counts")
# 
# log_info(note = "Filtering based on MT counts",
#          "Kept %d cells\nKept %d genes\n",
#          ncol(Seurat_filtered), nrow(Seurat_filtered))

```

```{r Step 3. Background and negative probe quality control}

Meta <- Seurat_filtered@meta.data %>%
  as.data.frame() 

Meta <- Meta %>%
  ## Make sure all are numeric ##
  mutate(across(c(propNegative, nCount_negprobes, nCount_falsecode, Total_Counts),
                ~ suppressWarnings(as.numeric(.))),
         ## Compute background ratios ##
         prop_negprobes_cnt = nCount_negprobes / Total_Counts,
         prop_falsecode_cnt = nCount_falsecode / Total_Counts,
         ## Final background metric ##
         propNegative = coalesce(propNegative, prop_negprobes_cnt))

## Visualise the negative probs ##
bg_plots <- list()

if ("propNegative" %in% colnames(Seurat_filtered@meta.data)) {
  bg_plots[["propNeg"]] <- VlnPlot(Seurat_filtered, features = "propNegative",
                                   group.by = "fov", pt.size = 0) + 
    ggtitle("Proportion of Negative Probe Signal by FOV")
}

if ("prop_falsecode_cnt" %in% colnames(Seurat_filtered@meta.data)) {
  bg_plots[["false"]] <- VlnPlot(Seurat_filtered, features = "prop_falsecode_cnt",
                                 group.by = "fov", pt.size = 0) + 
    ggtitle("Proportion of False Code Signal by FOV")
}

wrap_plots(bg_plots)

## Stat summary per group ##
Seurat_filtered@meta.data %>%
  group_by(Group) %>%
  summarise(n_cells = n(), 
            median_neg = median(propNegative, na.rm = TRUE),
            median_false = median(prop_falsecode_cnt, na.rm = TRUE),
            median_counts = median(Total_Counts, na.rm = TRUE), .groups = "drop") %>%
  arrange(desc(median_neg)) 

Seurat_filtered <- subset(Seurat_filtered, subset = (is.na(propNegative) | propNegative < 0.05) &
                            (is.na(prop_falsecode_cnt) | prop_falsecode_cnt < 0.05))

summary(Seurat_filtered@meta.data$propNegative)
summary(Seurat_filtered@meta.data$prop_falsecode_cnt)

Meta <- Meta %>%
  group_by(Group) %>%
  mutate(thr_neg = quantile(propNegative, 0.99, na.rm = TRUE),
         thr_fc = quantile(prop_falsecode_cnt, 0.99, na.rm = TRUE)) %>%
  ungroup() %>% 
  mutate(thr_neg = ifelse(is.finite(thr_neg), thr_neg, 0.05),
         thr_fc = ifelse(is.finite(thr_fc), thr_fc, 0.05))

keep_bg <- with(Meta, (is.na(propNegative) | propNegative < thr_neg) &
                      (is.na(prop_falsecode_cnt) | prop_falsecode_cnt < thr_fc))

cat("Will keep", sum(keep_bg, na.rm = TRUE), "of", ncol(Seurat_filtered), "cells\n")
stopifnot(sum(keep_bg, na.rm = TRUE) > 0)

Meta <- Meta %>% 
  column_to_rownames(var = "cell_ID")

Seurat_filtered@meta.data <- Meta

cells_keep <- colnames(Seurat_filtered)[keep_bg]
Seurat_filtered <- subset(Seurat_filtered, cells = cells_keep)

## Re filter the genes in case some of the removed resulted in counts of 0 ##
keep_genes <- GetAssayData(Seurat_filtered, assay = "RNA", layer = "counts") %>%
  { rowSums(. > 0) } %>% 
  enframe(name = "Gene", value = "nonzero") %>%
  filter(nonzero > 0) %>%
  pull(Gene)

rna_assay <- Seurat_filtered[["RNA"]]
rna_assay <- subset(rna_assay, features = keep_genes)
Seurat_filtered[["RNA"]] <- rna_assay

VlnPlot(Seurat_filtered, features = intersect(c("propNegative", "prop_falsecode_cnt"), 
                                             colnames(Seurat_filtered@meta.data)), pt.size = 0)

## Save Rds object to continue from here next session ##
SaveSeuratRds(object = Seurat_filtered, file = file.path(main_dir, "Data", "Edited_data", "Seurat_filtered_step_3.Rds"))

```

Visualisation based on the negative probs used in CosMx. If the following plots present a high number of negative probs being expressed apply additional filtering to remove cells or FOVs in extreme cases that are low quality based on these reads. 

```{r Clear space}

# keep <- c("Seurat_filtered", "log_file", "metrics", "qc_dir", "main_dir")
# rm(list = setdiff(ls(envir = .GlobalEnv), keep), envir = .GlobalEnv)
# gc()

```

```{r Visual QC based on negative probs}

# Seurat_filtered <- readRDS(file = paste0(main_dir, "Data/Edited_data/Seurat_filtered_step_3.Rds"))
Metadata_filtered <- Seurat_filtered@meta.data

xr <- range(Metadata_filtered$x_slide_mm, na.rm = TRUE)
yr <- range(Metadata_filtered$y_slide_mm, na.rm = TRUE)

summary(Seurat_filtered@meta.data$x_slide_mm)
summary(Seurat_filtered@meta.data$y_slide_mm)
length(unique(Seurat_filtered@meta.data$fov))

Seurat_filtered@meta.data %>%
  group_by(fov) %>%
  summarise(xmin = min(x_slide_mm, na.rm = TRUE),
            xmax = max(x_slide_mm, na.rm = TRUE),
            ymin = min(y_slide_mm, na.rm = TRUE),
            ymax = max(y_slide_mm, na.rm = TRUE))

## Total counts in space ##
g_p2 <- ggplot(Metadata_filtered, aes(x = x_slide_mm, y = y_slide_mm, color = Total_Counts)) +
  geom_point(size = 0.1, alpha = 0.8) +
  coord_fixed(xlim = xr, ylim = yr) +
  scale_color_viridis_c() +
  theme_void() + 
  ggtitle("Total counts (per cell)")

## Number of features in space ## 
g_p3 <- ggplot(Metadata_filtered, aes(x = x_slide_mm, y = y_slide_mm, color = Num_Features)) +
  geom_point(size = 0.1, alpha = 0.8) +
  coord_fixed(xlim = xr, ylim = yr) +
  scale_color_viridis_c() +
  theme_void() + 
  ggtitle("Features detected (per cell)")

## Background fractions in space (if present) ##
if ("propNegative" %in% names(Metadata_filtered)) {
  g_p4 <- ggplot(Metadata_filtered, aes(x = x_slide_mm, y = y_slide_mm, color = propNegative)) +
    geom_point(size = 0.1, alpha = 0.8) +
    coord_fixed(xlim = xr, ylim = yr) +
    scale_color_viridis_c(limits = c(0, quantile(Metadata_filtered$propNegative, 0.99, na.rm = TRUE))) +
    theme_void() + 
    ggtitle("Background (propNegative)")
}

if ("prop_falsecode_cnt" %in% names(Metadata_filtered)) {
  g_p5 <- ggplot(Metadata_filtered, aes(x = x_slide_mm, y = y_slide_mm, color = prop_falsecode_cnt)) +
    geom_point(size = 0.1, alpha = 0.8) +
    coord_fixed(xlim = xr, ylim = yr) +
    scale_color_viridis_c(limits = c(0, quantile(Metadata_filtered$prop_falsecode_cnt, 0.99, na.rm = TRUE))) +
    theme_void() + 
    ggtitle("Background (false code fraction)")
}

ggsave(filename = file.path(qc_dir, "Post_Filtering/FOV_QC_plots_Total_Counts_per_cell.png"),
       plot = g_p2, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "Post_Filtering/FOV_QC_plots_num_Features_per_cell.png"),
       plot = g_p3, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "Post_Filtering/FOV_QC_plots_propNegative_per_cell.png"),
       plot = g_p4, width = 9, height = 4, dpi = 300)

ggsave(filename = file.path(qc_dir, "Post_Filtering/FOV_QC_plots_False_code_Fraction_per_cell.png"),
       plot = g_p5, width = 9, height = 4, dpi = 300)

```

```{r Clear space}

# keep <- c("Seurat_filtered", "log_file", "metrics", "qc_dir", "main_dir")
# rm(list = setdiff(ls(envir = .GlobalEnv), keep), envir = .GlobalEnv)
# gc()

```

```{r Step 4. Filtering based off tissue outliers}

# Seurat_filtered <- readRDS(file = paste0(main_dir, "Data/Edited_data/Seurat_filtered_step_3.Rds"))

# Seurat_filtered <- readRDS(file = paste0(main_dir, "Data/Edited_data", "Seurat_filtered_step_3.Rds"))
Metadata_filtered <- Seurat_filtered@meta.data %>%
  as.data.frame() %>%
  rownames_to_column("cell_ID")

## Get the 95% quantiles to detect "outliers" ##
Metadata_filtered <- Metadata_filtered %>%
  group_by(Sample) %>%
  mutate(x_low = quantile(x_slide_mm, 0.005, na.rm = TRUE),
         x_high = quantile(x_slide_mm, 0.995, na.rm = TRUE),
         y_low = quantile(y_slide_mm, 0.005, na.rm = TRUE),
         y_high = quantile(y_slide_mm, 0.995, na.rm = TRUE)) %>%
  ungroup()

## Detect which cells are outliers based on the posiiton ##
keep_on_tissue <- with(Metadata_filtered, 
                       x_slide_mm >= x_low & x_slide_mm <= x_high &
                       y_slide_mm >= y_low & y_slide_mm <= y_high)

## Statistics to report the removed cell number ##
n_before <- ncol(Seurat_filtered)
n_keep <- sum(keep_on_tissue, na.rm = TRUE)
n_drop <- n_before - n_keep
pct_drop <- round((100 * n_drop / n_before), 2)

## Perform the filtering ##
Seurat_filtered <- subset(Seurat_filtered, cells = Metadata_filtered$cell_ID[keep_on_tissue])

SaveSeuratRds(object = Seurat_filtered, file = file.path(main_dir, "Data", "Edited_data", "Seurat_filtered_step_4.Rds"))

```

```{r Clear space}

# keep <- c("Seurat_filtered", "log_file", "metrics", "qc_dir", "main_dir")
# rm(list = setdiff(ls(envir = .GlobalEnv), keep), envir = .GlobalEnv)
# gc()

```

```{r}

## Pull the counts ##
Counts_filtered <- LayerData(Seurat_filtered, assay = "RNA", layer = "counts")

## Retrieve total expression after filtering ##
total_expression <- Matrix::rowSums(Counts_filtered)
## Retrieve number of cells each gene is expressed in ##
num_cells_expr <- Matrix::rowSums(Counts_filtered > 0)

## Create a small dataframe with these ##
Counts_stats <- tibble::tibble(gene = rownames(Counts_filtered),
                               total_counts = as.numeric(total_expression),
                               cells_detected = as.numeric(num_cells_expr),
                               ## fraction of cells with gene > 0 ##
                               prevalence = cells_detected / ncol(Counts_filtered)) %>% 
  mutate(prevalence_label = paste0(cells_detected, " (", round(prevalence * 100, 1), "%)"))

med_prev <- median(Counts_stats$prevalence, na.rm = TRUE)
# x_breaks <- sort(unique(c(seq(0, 1, by = 0.25), med_prev)))
x_breaks <- sort(unique(c(med_prev, 0.25, 0.5, 0.75, 1)))

## Gene level abundance (total transcripts per gene) ##
Hist_total_exp <- ggplot(Counts_stats, aes(x = total_counts)) +
  geom_histogram(bins = 60, color = "black", fill = "lightblue") +
  scale_x_log10() +
  geom_vline(xintercept = stats::median(Counts_stats$total_counts), linetype = 2) +
  labs(title = "Gene level abundance across CosMx cells",
       x = "Total transcripts per gene (log10 counts)",
       y = "Number of genes")

## Gene detection prevalence (cells detected per gene) ##
Hist_prev <- ggplot(Counts_stats, aes(x = prevalence)) +
  geom_histogram(bins = 60, color = "black", fill = "pink") +
  geom_vline(xintercept = stats::median(Counts_stats$prevalence), linetype = 2) +
  # scale_x_continuous(limits = c(0, 1),
  #                    breaks = x_breaks,
  #                    labels = lbl_cells_percent) +
  labs(title = "Gene detection prevalence (fraction of cells)",
       x = "Percentage of cells",
       y = "Number of genes")

g_p6 <- Hist_total_exp + Hist_prev  

ggsave(filename = file.path(qc_dir, "Post_Filtering/Counts_after_filtering_step_4.png"),
       plot = g_p6, width = 17, height = 4, dpi = 300)

```
